src/util/sighash.rs

```diff
@@ -101,6 +101,52 @@ pub struct ScriptPath<'s> {
     leaf_version: LeafVersion,
 }
 
+/// Script path spending details for computing the sighash of a Taproot tx
+pub enum TaprootSighashScriptPath<'s> {
+    /// Full script path, containing the script, leaf version and `OP_CODESEPARATOR` position
+    ScriptPath(ScriptPath<'s>),
+    /// Pre-computed leaf hash and `OP_CODESEPARATOR` position
+    LeafHash {
+        /// The leaf hash
+        leaf_hash: TapLeafHash,
+        /// The `OP_CODESEPARATOR` position
+        code_separator_pos: u32,
+    },
+}
+
+impl<'s> TaprootSighashScriptPath<'s> {
+    /// Return the leaf hash and `OP_CODESEPARATOR` position
+    pub fn leaf_hash_and_code_separator(&self) -> Result<(TapLeafHash, u32), Error> {
+        match self {
+            TaprootSighashScriptPath::ScriptPath(ScriptPath { script, code_separator_pos, leaf_version }) => {
+                let mut enc = TapLeafHash::engine();
+                leaf_version.as_u8().consensus_encode(&mut enc)?;
+                script.consensus_encode(&mut enc)?;
+
+                Ok((TapLeafHash::from_engine(enc), *code_separator_pos))
+            },
+            TaprootSighashScriptPath::LeafHash { leaf_hash, code_separator_pos } => Ok((*leaf_hash, *code_separator_pos)),
```

Comment:
I don't think this line is exercised by tests, is it possible to add one test passing here?